### ISSUE TITLE
Doesn't work with next.js Edge runtime

### ISSUE THREAD

**Issue by etodanik** [brianc/node-postgres#3206]
I'm writing an app on `next.js` which I plan to deploy to CloudFlare Pages + Workers. I know there's CloudFlare support, but when running the local development server of next.js, I get the following errors:
```
Learn More: https://nextjs.org/docs/messages/node-module-in-edge-runtime
    at <unknown> (webpack-internal:///(middleware)/./node_modules/next/dist/esm/server/web/globals.js:33)
    at Object.get (webpack-internal:///(middleware)/./node_modules/next/dist/esm/server/web/globals.js:33:19)
    at eval (webpack-internal:///(middleware)/./node_modules/pg/lib/crypto/utils-webcrypto.js:19:36)
    at (middleware)/./node_modules/pg/lib/crypto/utils-webcrypto.js (file:///home/danny/src/project/.next/server/pages/api/manageUser.js:3097:1)
    at __webpack_require__ (file:///home/danny/src/project/.next/server/edge-runtime-webpack.js:37:33)
    at fn (file:///home/danny/src/project/.next/server/edge-runtime-webpack.js:286:21)
    at eval (webpack-internal:///(middleware)/./node_modules/pg/lib/crypto/utils.js:7:22)
    at (middleware)/./node_modules/pg/lib/crypto/utils.js (file:///home/danny/src/project/.next/server/pages/api/manageUser.js:3108:1)
    at __webpack_require__ (file:///home/danny/src/project/.next/server/edge-runtime-webpack.js:37:33)
    at fn (file:///home/danny/src/project/.next/server/edge-runtime-webpack.js:286:21)
    at eval (webpack-internal:///(middleware)/./node_modules/pg/lib/crypto/sasl.js:3:16)
```

Would it make sense to generalize CloudFlare Workers detection into a broader Edge detection that also includes node edge runtime?

**Comment by brianc** [comment]
Yeah it would probably...how does one reproduce this issue locally?  If running in node then `pg` should be using the node pieces.

**Comment by etodanik** [comment]
Set up a basic `next.js` app, create an API endpoint, set its' runtime to edge by specifying:
`export const runtime = 'edge';`

And then assume to use `node-postgres` from within that route.

CloudFlare Pages requires a next.js app to specify that all backend endpoints will be using the edge runtime.

There's a further complication that `NodeEdge` runtime isn't the same as CloudFlare workers. I suppose next.js' NodeEdge is emulating what Vercel's edge runtime would look like. But one important detail is that it won't have the ability to run `cloudflare:sockets`

**Comment by muhammedaksam** [comment]
I am trying to use Prisma, PrismaPg and Pg.

> Error: The edge runtime does not support Node.js 'crypto' module.

**Call Stack**
```
Next.js
eval
node_modules\.pnpm\pg@8.11.5\node_modules\pg\lib\crypto\utils-webcrypto.js (21:1)
(middleware)/./node_modules/.pnpm/pg@8.11.5/node_modules/pg/lib/crypto/utils-webcrypto.js
file:///C:/Code/React/nextplate/.next/server/middleware.js (1292:1)
```
```
Next.js
eval
node_modules\.pnpm\pg@8.11.5\node_modules\pg\lib\crypto\utils.js (8:3)
(middleware)/./node_modules/.pnpm/pg@8.11.5/node_modules/pg/lib/crypto/utils.js
file:///C:/Code/React/nextplate/.next/server/middleware.js (1303:1)
```
```
Next.js
eval
node_modules\.pnpm\pg@8.11.5\node_modules\pg\lib\crypto\sasl.js (2:16)
```

https://vercel.com/docs/functions/runtimes/edge-runtime#unsupported-apis

I know it's not really an issue related to Pg but I wanted to inform here anyway.

**Comment by janpio** [comment]
Quick note from @prisma that this affects some of our users and is being reported to us: https://github.com/prisma/prisma/issues/24430 

(This issue has a reproduction, but that includes Prisma, Next.js and next-auth - that is probably a bit more than what you are asking for @brianc. I am pretty sure just trying to use `pg` in a Next.js middleware should have the same result though.)

We would love to see this fixed and released, so that Prisma users wanting to use `pg` in Next.js projects with middleware can do so.

**Comment by brianc** [comment]
k i'll get this looked at today if I have time!

**Comment by brianc** [comment]
@janpio @andyjy @etodanik would one of y'all be able to provide me w/ more concrete steps to follow to set up a local reproduction? I did `npx create-next-app@latest` and did NOT use the API router (so the legacy or older form of nextjs).  It created a hello api route:

<img width="421" alt="image" src="https://github.com/brianc/node-postgres/assets/50081/a5390203-8ec4-45cd-93d2-929d82ea8ef0">


So i tested it at http://localhost:3000/api/hello and it worked. great. then I went into the route file and added 

```ts
export const runtime = 'edge'
```

and now its erroring when I hit the route: 

<img width="1134" alt="image" src="https://github.com/brianc/node-postgres/assets/50081/5a354576-b340-4a03-861f-dbf50fb08c3d">


So I'm doing something wrong, but not sure what.

**Comment by brianc** [comment]
update: I tried a different next-js app using the `app router` instead of the old `pages/*` approach

I just did same steps as above except said "yes" when the setup asked me to use an app router.

Setup looks like this:

<img width="303" alt="image" src="https://github.com/brianc/node-postgres/assets/50081/b9964750-72f7-4967-b597-84ee55fafaea">

my route.ts file:

```ts
import * as pg from 'pg'

export const runtime = 'edge'
const pool = new pg.Pool()

export async function GET() {
  const { rows } = await pool.query('SELECT 1 as one')

  return Response.json({ rows })
}
```

Immediately when I run 'yarn next dev' to start the project it errors out with being unable to import `pg-native` which is annoying....I've never messed w/ the edge runtime before - it's quite a bit different...so any thoughts on this?


```

Import trace for requested module:
./node_modules/pg/lib/connection-parameters.js
./node_modules/pg/lib/client.js
./node_modules/pg/lib/index.js
./src/app/api/route.ts
./node_modules/next/dist/build/webpack/loaders/next-edge-app-route-loader/index.js?absolutePagePath=%2FUsers%2Fbmc%2Fsrc%2Fpg-edge-test%2Fsrc%2Fapp%2Fapi%2Froute.ts&page=%2Fapi%2Froute&appDirLoader=bmV4dC1hcHAtbG9hZGVyP25hbWU9YXBwJTJGYXBpJTJGcm91dGUmcGFnZT0lMkZhcGklMkZyb3V0ZSZhcHBQYXRocz0mcGFnZVBhdGg9cHJpdmF0ZS1uZXh0LWFwcC1kaXIlMkZhcGklMkZyb3V0ZS50cyZhcHBEaXI9JTJGVXNlcnMlMkZibWMlMkZzcmMlMkZwZy1lZGdlLXRlc3QlMkZzcmMlMkZhcHAmcGFnZUV4dGVuc2lvbnM9dHN4JnBhZ2VFeHRlbnNpb25zPXRzJnBhZ2VFeHRlbnNpb25zPWpzeCZwYWdlRXh0ZW5zaW9ucz1qcyZyb290RGlyPSUyRlVzZXJzJTJGYm1jJTJGc3JjJTJGcGctZWRnZS10ZXN0JmlzRGV2PXRydWUmdHNjb25maWdQYXRoPXRzY29uZmlnLmpzb24mYmFzZVBhdGg9JmFzc2V0UHJlZml4PSZuZXh0Q29uZmlnT3V0cHV0PSZwcmVmZXJyZWRSZWdpb249Jm1pZGRsZXdhcmVDb25maWc9ZTMwJTNEIQ%3D%3D&nextConfigOutput=&preferredRegion=&middlewareConfig=e30%3D!
 ○ Compiling /api ...
 ⚠ ./node_modules/pg/lib/native/client.js
Module not found: Can't resolve 'pg-native' in '/Users/bmc/src/pg-edge-test/node_modules/pg/lib/native'
```

I know its trying to webpack the server...which is _fun_.

**Comment by brianc** [comment]
> here you go - reproduction for pages router:

amazing, ty

**Comment by brianc** [comment]
@andyjy when i applied your patch to the crypto file I now get this:

```
⨯ Error: The edge runtime does not support Node.js 'net' module.
Learn More: https://nextjs.org/docs/messages/node-module-in-edge-runtime
    at Object.construct (webpack-internal:///(middleware)/./node_modules/next/dist/esm/server/web/globals.js:37:19)
    at getStream (webpack-internal:///(middleware)/./node_modules/pg/lib/stream.js:8:12)
    at new Connection (webpack-internal:///(middleware)/./node_modules/pg/lib/connection.js:19:36)
    at new Client (webpack-internal:///(middleware)/./node_modules/pg/lib/client.js:48:7)
    at handler (webpack-internal:///(middleware)/./src/pages/api/hello.ts:13:16)
    at eval (webpack-internal:///(middleware)/./node_modules/next/dist/esm/server/api-utils/index.js:32:16)
    at eval (webpack-internal:///(middleware)/./node_modules/next/dist/esm/server/lib/trace/tracer.js:115:36)
    at NoopContextManager.with (webpack-internal:///(middleware)/./node_modules/next/dist/compiled/@opentelemetry/api/index.js:2:7062)
    at ContextAPI.with (webpack-internal:///(middleware)/./node_modules/next/dist/compiled/@opentelemetry/api/index.js:2:518)
    at NoopTracer.startActiveSpan (webpack-internal:///(middleware)/./node_modules/next/dist/compiled/@opentelemetry/api/index.js:2:18108)
 ✓ Compiled in 107ms (263 modules)
```

**Comment by andyjy** [comment]
> @andyjy when i applied your patch to the crypto file I now get this:

ah ok, turns out I am using `import { Pool } from "pg"` in my app, which the patch above solves for - whereas in the reproduction repo I happened to write `import { Client } from "pg"` which exposes the new `net` incompatibility. 

Will keep digging..

**Comment by janpio** [comment]
Tiny PR to @andyjy's repro to show the same problem with a Next.js middleware, the common case where Prisma users run into this: https://github.com/andyjy/mre-pg-nextjs-edge/pull/1

> Thinking ahead to whether/how it's possible to get things actually working under the Next.js Edge runtime locally - it would probably(?) require breaking out of Next.js' local Edge Runtime "sandbox" to access node:net locally (i.e. via the Next.js Webpack config). So if this is the case, some tweaks to pg may be necessary but not sufficient alone.

Semi informed comment from the sidelines: Afaik the Cloudflare Workers support for `pg` had similar challenges, that parts of the internal `pg` code had to be disabled or skipped to not use APIs that are not available in the CfW environment.

**Comment by brianc** [comment]
> Sorry for the incorrect/misleading post previously!

absolutely no problem! I'll continue to noodle on this....might take a bit of time to do a similar retrofit that the CF stuff did. I'm also about to go out of town for a bit so...hang tight - i'll get there eventually!

**Comment by janpio** [comment]
🤔 
_Can_ `node-postgres` even work in the edge runtime by Vercel and Next.js? Isn't it missing TCP socket support, and Vercel does not offer a [`connect()` like Cloudflare](https://blog.cloudflare.com/workers-tcp-socket-api-connect-databases) does, that could be used instead.

That would be a limitation of Next.js functionality that uses the edge runtime (middleware, app routes) to not be able to support `pg` at all.

**Comment by andyjy** [comment]
> 🤔 _Can_ `node-postgres` even work in the edge runtime by Vercel and Next.js? Isn't it missing TCP socket support, and Vercel does not offer a [`connect()` like Cloudflare](https://blog.cloudflare.com/workers-tcp-socket-api-connect-databases) does, that could be used instead. That would be a limitation of Next.js functionality that uses the edge runtime (middleware, app routes) to not be able to support `pg` at all.

Yes, having looked into it more also - I agree it's essentially correct that lack of TCP socket support is a blocker for the Next.js/Vercel Edge Runtime.

Despite running on Cloudflare Workers behind the scenes(?) - when deploying to Vercel, code that references `require("cloudflare:sockets")` fails to deploy, with error `The Edge Function "middleware" is referencing unsupported modules: cloudflare:sockets`.

What I assume _is_ possible is deploying a Next.js app using the Edge runtime [directly to Cloudflare](https://developers.cloudflare.com/pages/framework-guides/nextjs/deploy-a-nextjs-site/). With no Vercel / Edge Runtime wrapper, the bundled code should just work on Cloudflare Workers.

And if someone wishes to do that, it is _technically_ possible to:

1. Patch `pg` to compile/bundle successfully under Next.js' Webpack config. This [requires a bunch of patching](https://github.com/andyjy/mre-pg-nextjs-edge/tree/main/patches).. and [additional next.config.js config](https://github.com/andyjy/mre-pg-nextjs-edge/blob/main/next.config.mjs#L6).

2. Patch Next.js to successfully `pg` under the Edge runtime (middleware, Pages API routes + App router) under local development (`next dev`), by breaking through the local Edge sandbox to expose `net.Socket`(!). Requires a [patch](https://github.com/andyjy/mre-pg-nextjs-edge/blob/main/patches/next%2B14.2.3.patch) and [additional next.config.js config](https://github.com/andyjy/mre-pg-nextjs-edge/blob/main/next.config.mjs#L11).

Potentially handy if anyone  _really_ wants to develop using Next.js + `pg` under Edge before deploying to Cloudflare..  but I can't imagine this ever being supported by Next.js "officially" unless the Edge runtime is somehow updated to incorporate TCP socket support...(!?)

[The reproduction repo](https://github.com/andyjy/mre-pg-nextjs-edge) is now updated with a proof of concept for both (1) and (2). I haven't deployed to Cloudflare directly but see no reason why this wouldn't work (potentially with a bit more tweaking to get the output bundle correct)

/nerd-snipe over

**Comment by janpio** [comment]
Thanks for the validation of my thought, I was not sure if I was _really_ mixing things up now.

Note, that will probably make this a lot more confusing instead of helping, as it makes it more complex: For Cloudflare _Pages_ there is an additional package `next-on-pages` so one can deploy Next.js to Cloudflare Pages, and that has its own problem with `pg`: https://github.com/cloudflare/next-on-pages/issues/605 Not sure if the "Edge Runtime" (from Next.js and Vercel) is even involved there anymore though.

Probably means no easy way to use `pg` in Next.js middlewares though 😢

**Comment by andyjy** [comment]
@janpio here's [the reproduction repo](https://github.com/andyjy/mre-pg-nextjs-edge) deployed directly to cloudflare via @cloudflare/next-on-pages

.. it works(!) - connecting to a Neon database using `node-postgres` and `@prisma/adapter-pg`:

- Pages router API route: https://44d82908.mre-pg-nextjs-edge.pages.dev/api/hello
- App router: https://44d82908.mre-pg-nextjs-edge.pages.dev/api/hello-app
- Edge middleware (you contributed): https://44d82908.mre-pg-nextjs-edge.pages.dev/api/foo

The patches to node-postgres I mentioned above are what fixes https://github.com/cloudflare/next-on-pages/issues/605

🎉

**Comment by andyjy** [comment]
@janpio ..and here it is deployed to Cloudflare [with Prisma added](https://github.com/andyjy/mre-pg-nextjs-edge/commit/abe4a1f24c9e8ad39742ae56c51cb4eeddefb84b) (via `@prisma/adapter-pg`):

Pages router API route: https://d2ff1a42.mre-pg-nextjs-edge.pages.dev/api/hello-app
App router: https://d2ff1a42.mre-pg-nextjs-edge.pages.dev/api/hello-app
Edge middleware: https://d2ff1a42.mre-pg-nextjs-edge.pages.dev/api/foo

**Comment by janpio** [comment]
Ok wow, that is pretty wild @andyjy - amazing!

Do I understand correctly that you seem a way to make `pg` work on Cloudflare Pages via `next-on-pages` via some changes? 
But deploying to Vercel is still completely out of the picture, correct?

**Comment by andyjy** [comment]
> Do I understand correctly that you seem a way to make `pg` work on Cloudflare Pages via `next-on-pages` via some changes

Exactly! And specifically those changes are:

> 1. Patch `pg` to compile/bundle successfully under Next.js' Webpack config. This [requires a bunch of patching](https://github.com/andyjy/mre-pg-nextjs-edge/tree/main/patches).. and [additional next.config.js config](https://github.com/andyjy/mre-pg-nextjs-edge/blob/main/next.config.mjs#L6).

Re: Vercel: 

> But deploying to Vercel is still completely out of the picture, correct

Yes - at least, I observe that when pushing bundled code to Vercel that includes `require("cloudflare:sockets")`, Vercel's deployment process objects with error `The Edge Function "...." is referencing unsupported modules: cloudflare:sockets`

(It's the same error if I push any code referencing e.g. `require("this_module_does_not_exist")`)

As I understand it - Vercel's Edge runtime product specifically supports their Edge Runtime API that is designed to be runnable under various V8-powered environments - AFAIK they don't particualrly publicise the fact it's Cloudflare Workers under the hood (and could host their Vercel Edge functions elsewhere under non-Cloudflare at any time - for all I know they may do already!). 

So at this point while it may be worth asking a friendly contact at Vercel about the possibility of bypassing the `unsupported module: cloudflare:sockets` error and running this code on Vercel, even if there is some way to make it work today I still wouldn't be too optimistic about it not breaking in future. The only practical way I can imagine would be if Vercel add support for a TCP socket library to their official Edge Runtime spec.

**Comment by janpio** [comment]
Shared that workaround over at `next-on-pages`: https://github.com/cloudflare/next-on-pages/issues/605#issuecomment-2154450280

**Comment by brianc** [comment]
Hi! I'm back from my trip - will be looking into this tomorrow morning! Thanks for all the help & research...I'll dive in tomorrow with some 👀 and 🔨

**Comment by brianc** [comment]
ugg...reading up on [the edge runtime vm](https://edge-runtime.vercel.app/packages/vm) trying to come up w/ a way to write some test for this within the repo. A bit complicated. Definitely open to ideas & help on this or a PR w/ the patch (though yeah I looked at it, it needs a bit more massaging).  Thanks for the work so far @andyjy!

**Comment by andresgutgon** [comment]
Same issues here using Drizzle with node-postgres and next-auth which use as DB adapter drizzle/pg. What I did is to do a dummy next-auth adapter stripping DB 

```javascript
import { AUTH_MAX_AGE } from '$/auth/constants';
import Google from 'next-auth/providers/google'
import NextAuth from 'next-auth'
import { NextResponse } from 'next/server';
import { NextAuthRequest } from 'node_modules/next-auth/lib';

// Make here the dummny next-auth without PostgreSQL dependency
const { auth } = NextAuth({
  session: { strategy: 'jwt', maxAge: AUTH_MAX_AGE },
  providers: [Google]
})

export default auth((_req: NextAuthRequest) => {
  return NextResponse.next();
})

export const config = {
  matcher: ["/((?!api|_next/static|_next/image|favicon.ico).*)"],
}
```

Not sure if this works but at least in development the page load

**Comment by muhammedaksam** [comment]
> Same issues here using Drizzle with node-postgres and next-auth which use as DB adapter drizzle/pg. What I did is to do a dummy next-auth adapter stripping DB
> 
> ```js
> import { AUTH_MAX_AGE } from '$/auth/constants';
> import Google from 'next-auth/providers/google'
> import NextAuth from 'next-auth'
> import { NextResponse } from 'next/server';
> import { NextAuthRequest } from 'node_modules/next-auth/lib';
> 
> // Make here the dummny next-auth without PostgreSQL dependency
> const { auth } = NextAuth({
>   session: { strategy: 'jwt', maxAge: AUTH_MAX_AGE },
>   providers: [Google]
> })
> 
> export default auth((_req: NextAuthRequest) => {
>   return NextResponse.next();
> })
> 
> export const config = {
>   matcher: ["/((?!api|_next/static|_next/image|favicon.ico).*)"],
> }
> ```
> 
> Not sure if this works but at least in development the page load

If you are going to use JWS as strategy you probably don't need to use an adapter for your Next Auth so you probably don't need to use Pg either.

**Comment by andresgutgon** [comment]
> If you are going to use JWT as strategy you probably don't need to use an adapter for your Next Auth so you probably don't need to use Pg either.

I still need to use PG adapter for storing users and accounts

**Comment by vnce** [comment]
For those hitting this issue as part of a next-auth / auth js 5 integration, this document helped me understand the situation better: https://authjs.dev/guides/edge-compatibility

"""
PostgreSQL is a database that uses a message-based protocol for communication between a the client and server that is transported via TCP (or Unix) sockets. Raw TCP sockets are one of those Node.js features that are generally not available to edge runtimes. Therefore, on the surface, it seems like it’s not possible to communicate with a PostgreSQL database from JavaScript running on edge runtime. The same goes for many other databases and their respective communication protocols.

As edge runtimes have matured and become more popular, however, people have gotten creative and implemented various solutions to this problem. One such common solution is to put some sort of API server in front of the database whose goal is to translate database queries sent to it via HTTP into a protocol the database can understand. This allows the client side to only have to make HTTP requests to the API server, which is something that every edge runtime supports.
"""

My workaround was to expose an API to pg that an auth.ts callback can hit from edge. I think I'd still prefer using pg in edge runtime (sounds like Brian is close to a solution), but maybe this is working as intended if node-postgres is designed for the node.js runtime, which Vercel's Edge runtime doesn't (or only partially) uses

**Comment by andresgutgon** [comment]
> My workaround was to expose an API to pg that an auth.ts callback can hit from edge. I think I'd still prefer using pg in edge runtime (sounds like Brian is close to a solution), but maybe this is working as intended if node-postgres is designed for the node.js runtime, which Vercel's Edge runtime doesn't (or only partially) uses

But you still can have middleware if you use `jwt` and `node-postgresql`. [This worked for me](https://github.com/brianc/node-postgres/issues/3206#issuecomment-2186342503) and is explained in [the docs here](https://authjs.dev/guides/edge-compatibility#the-solution)

**Comment by Goldziher** [comment]
This is sufficent:

```typescript
import { MiddlewareConfig, NextResponse } from "next/server";
import { PagePath } from "@/enums";
import NextAuth from "next-auth";

const { auth } = NextAuth({ providers: [] });

export const middleware = auth(async (req) => {
	const path = new URL(req.url).pathname;

	const session = await auth();

	const isSigninPage = path === PagePath.SIGNIN.toString();
	const isRootPage = path === PagePath.ROOT.toString();
	const isAPIRoute = path.startsWith("/api");

	if (session && isSigninPage) {
		return NextResponse.redirect(new URL(PagePath.ROOT, req.url));
	}

	if (!session && !(isRootPage || isSigninPage || isAPIRoute)) {
		return NextResponse.redirect(new URL(PagePath.SIGNIN, req.url));
	}
});

export const config: MiddlewareConfig = {
	matcher: ["/((?!api|_next/static|_next/image\\.(?:svg|png|jpg|jpeg|gif|webp)$).*)"],
};
```

BUT you have to ensure that you are using the JWT strategy and never access the DB during the middleware, here is auth.ts:

```typescript
import NextAuth, { DefaultSession } from "next-auth";
import { DrizzleAdapter } from "@auth/drizzle-adapter";
import { getDatabaseClient } from "db/connection";
import { accounts, sessions, users, verificationTokens } from "db/schema";
import { User } from "@/types/database-types";
import { eq } from "drizzle-orm";
import Google from "next-auth/providers/google";
import Resend from "next-auth/providers/resend";

declare module "next-auth" {
	/**
	 * Returned by `auth`, `useSession`, `getSession` and received as a prop on the `SessionProvider` React Context
	 */
	interface Session {
		user: User & DefaultSession["user"];
	}
}

export const { handlers, signIn, signOut, auth } = NextAuth({
	adapter: DrizzleAdapter(getDatabaseClient(), {
		usersTable: users,
		accountsTable: accounts,
		sessionsTable: sessions,
		verificationTokensTable: verificationTokens,
	}),
	session: { strategy: "jwt" },
	callbacks: {
		session: async ({ session }) => {
			const db = getDatabaseClient();
			const user = await db.query.users.findFirst({
				where: eq(users.email, session.user.email),
			});

			session.user = { ...user, ...session.user };
			return session;
		},
	},
	providers: [Google, Resend],
});
```

**Comment by the-me-0** [comment]
Hello there !
I'm currently running in this issue while trying to get prisma to run in nextjs edge runtime. Their documentation is telling me to use something like this :
```ts
import { Pool } from 'pg';
import { PrismaPg } from '@prisma/adapter-pg';
import { PrismaClient } from '@prisma/client';

const pool = new Pool({ connectionString });
const adapter = new PrismaPg(pool);
const prisma = new PrismaClient({ adapter });
```

but as soon as I make a database call, I get this kind of error :
```
Module not found: Can't resolve 'fs'
https://nextjs.org/docs/messages/module-not-found

Import trace for requested module:
./node_modules/pg/lib/connection-parameters.js
./node_modules/pg/lib/client.js
./node_modules/pg/lib/index.js
./instrumentation.ts
```

As you can see, the error originates from the "pg" library. I'm gonna be completely honest and tell you that I don't really know what the "pg" is (although I guess it allows to connect to postgres databases without using TCP). I'm gonna try to create a minimal repo that recreates this behaviour asap, but reading this whole issue comments I did not see a solution come out yet.

Have a nice day, and do not hesitate to ping me if you manage to get results on this matter.

**Comment by the-me-0** [comment]
Minimal repro of my situation here : [https://github.com/the-me-0/node-pg-minimal-repro](https://github.com/the-me-0/node-pg-minimal-repro)
I'm still trying to resolve it on my side, and I'll keep you updated if I find anything.

**Comment by andyjy** [comment]
@the-me-0 `pg` isn't compatible with Next.js Edge Runtime (which == Vercel Edge Functions). Next.js doesn't support all Cloudflare's APIs - in particular the required `connect()`.

In order to use Prisma with Next.js Edge, you need to use a supported Prisma Driver Adapter/database combination here:
https://www.prisma.io/docs/orm/prisma-client/deployment/edge/overview#which-database-drivers-are-edge-compatible .

Note that page says:
> [node-postgres](https://node-postgres.com/) (pg) uses Cloudflare's connect() (TCP) to access the database. It is only compatible with Cloudflare Workers, not with Vercel Edge Functions.

In other words, `pg` _does_ work when used directly with Cloudflare Workers - but to do this you'd need to use a different framework, not Next.js. You could use something more lightweight like [Hono](https://hono.dev/) - or even no framework at all, as in the [Prisma guide for Cloudflare Workers](https://www.prisma.io/docs/orm/prisma-client/deployment/edge/deploy-to-cloudflare)

**Comment by the-me-0** [comment]
I did read this page yesterday, and I am aware that pg cannot run in nextjs's edge runtime. But this issue is still open, so isn't that a thing "pg" is aiming to fix ? If so, my repro could help. If not, this issue should be closed.

I needed pg to work in this edge runtime because no other solution really cut it for me (fortunately, I found a way around, documented on the prisma issue : [https://github.com/prisma/prisma/issues/24430#issuecomment-2471286836](https://github.com/prisma/prisma/issues/24430#issuecomment-2471286836)), as I have a local DB running on docker and that's what I'm trying to connect to. I'll admit that I did not try to use other drivers proposed by the prisma documentation, as I did not know if they would work for my "localhost" set-up.

Maybe they do and I just lost a lot of time for nothing, but in any case I got something to work now.

**Comment by mkpanq** [comment]
For all who uses drizzle ORM with `node-postgres` adapter - I've also had `Error: The edge runtime does not support Node.js 'crypto' module.`
My solution was to replace driver from `node-postgres` to `postgres.js` - more info about it [here](https://orm.drizzle.team/docs/get-started-postgresql#step-1---install-packages-1) 

Hope it's gonna be possibile in your projects !

**Comment by jvandenaardweg** [comment]
postgres.js also does not work:

```
../../node_modules/postgres/src/connection.js:2:1
Module not found: Can't resolve 'tls'
````

According to the drizzle docs there are only a few edge compatible drivers, node-postgres and postgres/postgres.js arent listed: https://orm.drizzle.team/docs/tutorials/drizzle-with-vercel-edge-functions
